### ISSUE TITLE
the parameter of the weight_behind?

### ISSUE THREAD

**Issue by zj19921221** [huawei-noah/Disout#4]
hi   i have some question about the code;
I have write the code in tensorflow yestoday; 
In my opinion , as the code released show that the disout is on the bisic of the dropblock added with a random value;
<0>in your code, the weight_behind is always set None ;
<1> i am also confused about the ERC after reading the paper; why use "dist=self.alpha*0.01*(var**0.5)*torch.randn(*x.shape,device=x.device)" to get the ERC? where mentioned in your paper?

**Comment by yehuitang** [comment]
Hi，for Q1, the self.weight_behind is not None in the training phase because of  the code 'modulelist[dploc[imodu]].weight_behind=modulelist[convloc[imodu]].weight.data' in resnet.py.  For Q2,  you can refer Equation (15) in the original paper, which is an appromiated gradient related to ERC.

**Comment by JackieTien97** [comment]
> For Q2, you can refer Equation (15) in the original paper, which is an appromiated gradient related to ERC.

You mean dist is the appromiated gradient, but where do you use the gradient to update the ξ, actually, I even didn't see the ξ in the code.

**Comment by zj19921221** [comment]
thanks for your reply!
I have another question about calculating the weight_mean;
in resnet all disout block is after a normal conv, the shape of the weight  is out_c , in_c, k_h, k_w;
but my backbone is mobilenet_v2, which is after a depth_wise conv , the shape of weight is out_c, 1, k_h, k_w; 
Can i use the same way to calculate the weight_mean;
have you ever use the disout in the depth_wise model like mobilenet_v  or shuffle_net??

**Comment by yehuitang** [comment]
We did not conduct experiments on depth-wise models. Theoretically, the form of distortion is similar.

**Comment by yehuitang** [comment]
@JackieTien97 (15) is the gradient  and in pactice we update \varepsilon one step in an iteration as described in experiments. The code in disout.py contains updating \varepsilon one step and applying \varepsilon on features.

**Comment by JackieTien97** [comment]
> The code in disout.py contains updating \varepsilon one step and applying \varepsilon on features.

@yehuitang Thank you for your explanation. But I am still confused about where you update the ξ.
If the variable `dist` in `disout.py` represents  ξ. The following code is calculating the dist:
```
if not (self.weight_behind is None) and not (len(self.weight_behind) == 0):
  dist = self.alpha * weight_mean * (var ** 0.5) * torch.randn(*x.shape, device=x.device)
else:
  dist = self.alpha * 0.01 * (var ** 0.5) * torch.randn(*x.shape, device=x.device)
```
But where did you update it, it seems that you didn't save the before value of `dist` and recalculate the `dist` in each train step which is not consistent with the progress in the paper you described.

**Comment by yehuitang** [comment]
At each iteration of weights updating, the initial distortion is set to  feature 'f(x)' and distortion is updated one step via gradient in practice. The 'dist'  in the code is  related to the grdient and step length.  This is consistent to the describition in the paper. You can refer to the section 'Optimization of the Distortion' and 'Experiments'.
